lib/trie/version.go

```diff
@@ -18,27 +27,71 @@ const (
 	// V0 is the state trie version 0 where the values of the keys are
 	// inserted into the trie directly.
 	// TODO set to iota once CI passes
-	V0 Version = 1
+	V0 Version = iota
+	V1
 )
 
+// ErrParseVersion is returned when parsing a state trie version fails.
+var ErrParseVersion = errors.New("parsing version failed")
+
+// DefaultStateVersion sets the state version we should use as default
+// See https://github.com/paritytech/substrate/blob/5e76587825b9a9d52d8cb02ba38828adf606157b/primitives/storage/src/lib.rs#L435-L439
+const DefaultStateVersion = V1
+
+// Entry is a key-value pair used to build a trie
+type Entry struct{ Key, Value []byte }
+
+// Entries is a list of entry used to build a trie
+type Entries []Entry
+
+// String returns a string representation of trie version
 func (v Version) String() string {
 	switch v {
 	case V0:
 		return "v0"
+	case V1:
+		return "v1"
 	default:
 		panic(fmt.Sprintf("unknown version %d", v))
 	}
 }
 
-var ErrParseVersion = errors.New("parsing version failed")
+// ShouldHashValue returns true if the value should be hashed based on trie version
+func (v Version) ShouldHashValue(value []byte) bool {
+	return v == V1 && len(value) > V1MaxValueSize
+}
+
+// Root returns the root hash of the trie built using the given entries
+func (v Version) Root(entries Entries) (common.Hash, error) {
```

Comment:
I would have assumed this method would be on the `Trie` and would accept a version as a parameter, similar to how the host api functions are versioned.

Reply:
Thats an option, but I prefer to use this one that is more like [substrate](https://github.com/paritytech/polkadot-sdk/blob/18257373b313fcfad4f7a2ab2cb02cb8bfc5b8f1/substrate/primitives/trie/src/lib.rs#L81)